genesis/recorders/file_writers.py

```diff
@@ -0,0 +1,351 @@
+            self.fps,
+            (self.width, self.height),
+            self.is_color,
+        )
+        if not video_writer.isOpened():
+            gs.raise_exception(f'[{type(self).__name__}] Failed to open video writer for "{filename}"')
+
+    def process(self, data, cur_time):
+        if self.is_color is None:
+            self._initialize_data(data)
+        if self.video_writer is None:
+            self._initialize_writer()
+
+        if self.is_color:
+            data = cv2.cvtColor(data, cv2.COLOR_RGB2BGR)
+        self.video_writer.write(data)
+
+    def cleanup(self):
+        if self.video_writer:
+            self.video_writer.release()
+            gs.logger.info(f'Video saved to "~<{self._options.filename}>~".')
+            self.video_writer = None
+
+    @property
+    def run_in_thread(self) -> bool:
+        return True
+
+
+class CSVFileWriterOptions(BaseFileWriterOptions):
+    """
+    Writes to a .csv file using `csv.writer`.
+
+    Can handle any array-like or dict[str, array-like] output, e.g. from sensors.
+
+    Parameters
+    ----------
+    filename : str
+        The name of the CSV file to save the data.
+    header : list[str] | None, optional
+        Column headers for the CSV file.
+    flush_interval : int | None, optional
+        Determines how often the data is saved to disk. None = only at cleanup, 0 = every write, >0 = every N writes.
+    save_on_reset: bool, optional
+        Whether to save the data on reset. Defaults to False.
+        If True, a counter will be added to the filename and incremented on each reset.
+    """
+
+    header: list[str] | None = None
+    flush_interval: int | None = None
+
+    def validate(self):
+        super().validate()
+        if not self.filename.lower().endswith(".csv"):
+            gs.raise_exception(f"[{type(self).__name__}] CSV output must be a .csv file")
+        if self.flush_interval is not None and self.flush_interval < 0:
+            gs.raise_exception(f"[{type(self).__name__}] flush_interval must be None or >= 0")
+
+
+@register_recording(CSVFileWriterOptions)
+class CSVFileWriter(BaseFileWriter):
+
+    def initialize(self):
+        super().initialize()
+        if self._options.flush_interval is not None:
+            self.flush_counter = 0
+
+        os.makedirs(os.path.abspath(os.path.dirname(self._options.filename)), exist_ok=True)
+        self._initialize_writer()
+
+    def _initialize_writer(self):
+        self.wrote_header = False
+        self.wrote_data = False
+        self.file_handle = open(self._get_filename(), "w", encoding="utf-8", newline="")
+        self.csv_writer = csv.writer(self.file_handle)
+
+    def process(self, data, cur_time):
+        row_data = [cur_time]
+        if isinstance(data, dict):
+            for value in data.values():
+                if isinstance(value, torch.Tensor):
+                    value = value.tolist()
+                row_data.append(value)
```

Comment:
You should probably assert that `value` is valid at this point. At least at first run.